File: schema/schema-ruby-erb/model_generation/Models/ServerDriven/Screen/safe_area.rb
#   Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA

#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
 
#       http://www.apache.org/licenses/LICENSE-2.0
 
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

require_relative '../../../Synthax/Attributes/variable.rb'
require_relative '../../base_component.rb'
require_relative '../../../Synthax/Types/built_in_type.rb'
require_relative '../server_driven_component.rb'
require_relative '../../Context/context_component.rb'

class SafeArea < BaseComponent

    def initialize
        variables = [
            Variable.new(
                :name => "top",
                :typeName => TypesToString.bool,
                :isOptional => true,
                :comment => "enable the safeArea constraint only on the TOP of the screen view."
            ),
            Variable.new(
                :name => "leading",
                :typeName => TypesToString.bool,
                :isOptional => true,
                :comment => "enable the safeArea constraint only on the LEFT side of the screen view."
            ),
            Variable.new(
                :name => "bottom",
                :typeName => TypesToString.bool,
                :isOptional => true,
                :comment => "enable the safeArea constraint only on the BOTTOM of the screen view."
            ),
            Variable.new(
                :name => "trailing",
                :typeName => TypesToString.bool,
                :isOptional => true,
                :comment => "enable the safeArea constraint only on the RIGHT of the screen view."
            ),
        ]
        synthax_type = BuiltInType.new(
            :name => self.name,
            :variables => variables,
            :package => "br.com.zup.beagle.widget.core",
            :comment => "The SafeArea will enable Safe areas to help you place your views within the visible portion of the overall interface. Note: This class is only used to iOS SafeArea"
        )

        super(synthax_type)

    end

end

File: iOS/Sources/Beagle/Sources/Components/ServerDrivenComponent/Screen/SafeArea.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/// The SafeArea will enable Safe areas to help you place your views within the visible portion of the overall interface.
public struct SafeArea: Equatable, Decodable, AutoInitiable {

    /// Enables the safeArea constraint only on the TOP of the screen view.
    public let top: Bool?
    
    /// Enables the safeArea constraint only on the LEFT side of the screen view.
    public let leading: Bool?
    
    /// Enables the safeArea constraint only on the BOTTOM of the screen view.
    public let bottom: Bool?
    
    /// Enables the safeArea constraint only on the RIGHT of the screen view.
    public let trailing: Bool?
    
// sourcery:inline:auto:SafeArea.Init
    public init(
        top: Bool? = nil,
        leading: Bool? = nil,
        bottom: Bool? = nil,
        trailing: Bool? = nil
    ) {
        self.top = top
        self.leading = leading
        self.bottom = bottom
        self.trailing = trailing
    }
// sourcery:end
}
